### DISCUSSION TITLE
Studio GPU acceleration

### DISCUSSION THREAD

**Discussion by zkeyg** [foxglove/community#1045] [Q&A]
Hi,
I would like to understand more about how to enable GPU acceleration for Foxglove. It appears that when we run foxglove on our machine (x86 linux ubuntu22.04, Nvidia RTX2000 series, Driver Version: 535), it has a high CPU usage whereas GPU usage remains unchanged.
We followed the doc to enable WebGL on Chrome but it seems it did not improve the performance. Is there a way to tell if the application is using GPU? And is there another other flag I can set for GPU acceleration?

Thanks!

**Comment by defunctzombie** [comment]
What do you see in chrome when you visit `chrome://gpu/` ?

**Reply by zkeyg** [reply]
Hi @defunctzombie 
Thanks for the reply! We used the official document and configure the chrome://gpu. It listed WebGL as hardware accelerated.  

*   Canvas: Hardware accelerated
*   Canvas out-of-process rasterization: Enabled
*   Direct Rendering Display Compositor: Disabled
*   Compositing: Hardware accelerated
*   Multiple Raster Threads: Enabled
*   OpenGL: Enabled
*   Rasterization: Hardware accelerated
*   Raw Draw: Disabled
*   Skia Graphite: Disabled
*   Video Decode: Hardware accelerated
*   Video Encode: Software only. Hardware acceleration disabled
*   Vulkan: Disabled
*   WebGL: Hardware accelerated
*   WebGL2: Hardware accelerated
*   WebGPU: Disabled
*   WebNN: Disabled

**Reply by defunctzombie** [reply]
So that looks like the right things are enabled. Do you experience lag using the app when trying to visualize items in the 3d panel? Is that why you are wondering whether GPU is being used? CPU usage is going to depend on what else you're visualizing in the app.

**Reply by zkeyg** [reply]
Yes @defunctzombie We experienced lagging while visulizing 3D Pointclouds and wanted to see if there is a way to make sure Foxglove is using GPU.

**Comment by tonynajjar** [comment]
Hello, I'm also experiencing performance issues whenever I'm publishing e.g 2000 Sphere/Arrow Primitive + 2000 TextPrimitive. 

![image](https://github.com/user-attachments/assets/8870c129-7f0e-4f5e-8abf-ee1fb0b93217)
![image](https://github.com/user-attachments/assets/e94504aa-8344-41c5-9077-1de6197fb672)


Both WebGL and WebGL2 are hardware accelerated as mentioned [here](https://docs.foxglove.dev/docs/visualization/panels/3d#performance-troubleshooting).

![image](https://github.com/user-attachments/assets/fff50738-4d99-411f-ab0a-00ee2db73ad9)
I can see that the second process (PID 49736) shows up in `nvidia-smi` but not the first one, which uses the most CPU. Is this expected?

My GPU is RTX 4060

**Reply by defunctzombie** [reply]
Thanks - I've ticketed this for us to investigate. If you can share this sample data it would be helpful for us to understand if its specific to a kind of data. We can make sample data but its not always an exact match for the data you've produced.

**Comment by nthiad** [comment]
I've attempted to replicate this performance issue and it could be related to how updates are written in the mcap.

Here are some scripts I used to generate mcap data of a similar scale (2000 arrow and text labels) where the fast version contains all the arrow and text primitives in a single SceneUpdate message per update and the slow version wraps each arrow and text primitive individually in a SceneUpdate:

https://gist.github.com/nthiad/2a13d2e1003814bca5e7e3052192e51d

If you likewise notice that fast.mcap performs well and slow.mcap performs poorly, then the problem may lie with how the data for SceneUpdates is formatted across scene primitives. If fast.mcap performs poorly, we can perhaps narrow in on some other cause but using a common benchmark.

Here are download links for the mcap files:

* [fast.mcap](https://drive.google.com/file/d/13ZbnxmafSr2MTcMsnvGl7_4R2OGXhGkP/view?usp=sharing) (70M)
* [slow.mcap](https://drive.google.com/file/d/1mCO_EjdYrTryDyyNxBmRd21siB6FJdJH/view?usp=sharing) (100M)

**Reply by tonynajjar** [reply]
> If you likewise notice that fast.mcap performs well and slow.mcap performs poorly,

I can confirm this observation

**Reply by nthiad** [reply]
Thanks! This confirms my theory for what's going on and the changes we can make in the rendering layer around SceneEntities that get drawn on the same frame to reduce the number of draw calls each frame. It could also help us to know more about your data model here and your use case to have a little more context to make sure these performance improvements fully address the slowdown you're seeing and the features in the data model you're using, if you can share.

**Reply by tonynajjar** [reply]
@nthiad I thought you were another user. I understand right that you work at Foxglove? I can definitely try to share a bag within this week. I've never yet tested it with a bag, only live visualizations but I would be surprised if it doesnt reproduce. Will keep you posted!

**Reply by nthiad** [reply]
Yep! I work on the performance / low level team for frontend so it's my job to look into these things.

>  I can definitely try to share a bag within this week. I've never yet tested it with a bag, only live visualizations but I would be surprised if it doesnt reproduce. Will keep you posted!

That would be great, thanks! If you don't want to share publically here in this thread you can send them to synthia@foxglove.dev

**Reply by tonynajjar** [reply]
I found some time now, sent it to your email. Thanks for taking a look!

**Reply by nthiad** [reply]
Just to keep you posted, I made a prototype change to how we render SceneEntities last week which renders this test data much faster. It needs more polish to track some information in different places  but I should hopefully get some time to finish this performance improvement soon.

**Reply by tonynajjar** [reply]
Hello @nthiad, do you have any update for me? Thank you

**Reply by nthiad** [reply]
I started working on the second pass on this issue on Friday to overhaul rendering so that instancing happens across all scene entities, not only a single scene entity. This week we are busy with an internal hack week but Monday I'm back on that ticket and will hopefully be able to ship something next week. In the meantime as an immediate workaround, if you can group your arrow and text geometry into a single scene entity it will speed things up a lot.

**Reply by tonynajjar** [reply]
Hello! can you give a quick status update here? Thanks

**Reply by nthiad** [reply]
Hi! I have a PR for this feature that we're had a lot of iteration on and are pretty close to landing. I had to overhaul a lot of how the instancing and rendering hierarchy works which took a while, but things are looking good for hopefully landing it this week and if that happens it will ship in the next release. There are still some minor issues with the render order in the more performant instancing mode compared to what we had before that our visual test suite is picking up but those are the only open issue left to resolve.

**Comment by nthiad** [comment]
I've just merged a big improvement for this case into main, so you should be able to test this out on web now and it may or may not be in this desktop release (cutting it close to the wire on the new release) but if not it will be in the next one. Thanks for your patience, this one was a doozy of a refactor but puts us in a good spot to speed up all the other scene entity primitives following on from this work.

I'm also working on a docs PR to discuss the performance impacts of different SceneUpdate patterns including the improvements specific to arrows we've rolled out.

**Reply by tonynajjar** [reply]
Thank you so much, will give it a try soon!

**Comment by evTessellate** [comment]
Hi! I also have an RTX2000 series and I can't seem to use it for Foxglove. I'm on a Ubuntu 20.04 os. From my debugging with ChatGPT, I found out that I couldn't force it to use GLX (which could be the solution to my problem), because the desktop app is not build to support GLX.

ChatGPT suggest that building the version of the APP with GLX enable could work. Is it possible to provide such build so I can test it?

Here's my command to launch foxglove with the current error log:
_NV_PRIME_RENDER_OFFLOAD=1 __GLX_VENDOR_LIBRARY_NAME=nvidia \ \ foxglove-studio --use-gl=desktop --ignore-gpu-blocklist --enable-gpu-rasterization --enable-zero-copy Could not set app as handler for foxglove:// [70166:0818/111805.101661:ERROR:ui/gl/init/gl_factory.cc:103] Requested GL implementation (gl=none,angle=none) not found in allowed implementations: [(gl=egl-angle,angle=default)]. [70166:0818/111805.102338:ERROR:components/viz/service/main/viz_main_impl.cc:184] Exiting GPU process due to errors during initialization Checking for beta autoupdate feature for deb/rpm distributions Found package-type: deb [70202:0818/111805.205133:ERROR:ui/gl/init/gl_factory.cc:103] Requested GL implementation (gl=none,angle=none) not found in allowed implementations: [(gl=egl-angle,angle=default)]. [70202:0818/111805.206044:ERROR:components/viz/service/main/viz_main_impl.cc:184] Exiting GPU process due to errors during initialization libva error: /usr/lib/x86_64-linux-gnu/dri/iHD_drv_video.so init failed libva error: /usr/lib/x86_64-linux-gnu/dri/i965_drv_video.so init failed

**Comment by xhubbard-swan** [comment]
Having similar issues! On the desktop application I get the warning about hardware acceleration not being enabled, but not so on chrome. 

Using a Dell Precision 5690, Intel Integrated Graphics, 32GB
